compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -112,10 +112,15 @@ object RefChecks {
    *  a class or module with same name
    */
   private def checkCompanionNameClashes(cls: Symbol)(implicit ctx: Context): Unit =
-    if (!(cls.owner is ModuleClass)) {
-      val other = cls.owner.linkedClass.info.decl(cls.name).symbol
-      if (other.isClass)
+    if (!cls.owner.is(ModuleClass)) {
+      def clashes(sym: Symbol) =
+        sym.name.stripModuleClassSuffix == cls.name.stripModuleClassSuffix &&
+        sym.isClass
```

Comment:
I would reverse the order of the conditions since `sym.isClass` is cheaper to compute